source/simulator/nullspace.cc

```diff
@@ -19,22 +19,164 @@
+    {
+      const auto &fe = dof_handler.get_fe();
+      const std::vector<Point<dim - 1>> &unit_support_points = fe.get_unit_face_support_points();
+      const Quadrature<dim - 1> quadrature(unit_support_points);
+      const unsigned int dofs_per_face = fe.dofs_per_face;
+      std::vector<types::global_dof_index> face_dofs(dofs_per_face);
+
+      FEFaceValues<dim, spacedim> fe_face_values(mapping,
+                                                 fe,
+                                                 quadrature,
+                                                 update_quadrature_points);
+
+      for (const auto &cell : dof_handler.active_cell_iterators())
+        if (cell->at_boundary()
+            &&
+            (cell->is_locally_owned() || cell->is_ghost()))
+          for (unsigned int face_no = 0;
+               face_no < GeometryInfo<dim>::faces_per_cell; ++face_no)
+            if (cell->at_boundary(face_no))
+              {
+                typename DoFHandler<dim, spacedim>::face_iterator face = cell->face(face_no);
+                face->get_dof_indices(face_dofs);
+                fe_face_values.reinit(cell, face_no);
+
+                bool found = false;
+                for (unsigned int i = 0; i < face_dofs.size(); ++i)
+                  {
+                    const unsigned int component = fe.face_system_to_component_index(i).first;
+                    if (mask[component])
+                      {
+                        const Point<dim> position = fe_face_values.quadrature_point(i);
+                        if (position.distance(location) < 1e-6*cell->diameter())
+                          {
+                            found = true;
+                            if (!constraints.is_constrained(face_dofs[i]) &&
+                                constraints.can_store_line(face_dofs[i]))
+                              constraints.add_line(face_dofs[i]);
+                          }
+                      }
+                  }
+
+                // Success! No reason to look at any other faces:
+                if (found)
+                  return true;
+              }
+
+      return false;
+    }
+
+    /**
+     * Like above, but on the given multigrid level @p level.
+     */
+    template <int dim, int spacedim=dim>
+    bool
+    constrain_point_on_boundary_to_zero(const DoFHandler<dim,spacedim> &dof_handler,
+                                        const Mapping<dim> &mapping,
+                                        const Point<dim> &location,
+                                        const ComponentMask &mask,
+                                        const int level,
+                                        AffineConstraints<double> &constraints)
+    {
+      // Call the active one instead if level==-1. This indirection simplifies the code using this function.
+      if (level==-1)
+        return constrain_point_on_boundary_to_zero(dof_handler, mapping, location, mask, constraints);
```

Comment:
Well, I'm still not fond of the -1 thing :-( Do you insist on it?

Reply:
No, still working on it...